internal/plugin/common/models.go

```diff
@@ -0,0 +1,65 @@
+package common
+
+import (
+	"reflect"
+	"strings"
+)
+
+type Index int
+
+const (
+	ProjectIndex  Index = 0
+	ServiceIndex  Index = 1
+	DatabaseIndex Index = 2
+)
+
+func PathIndex(s string, i Index) string {
+	list := strings.Split(s, "/")
+	if int(i) < len(list) {
+		return list[i]
+	}
+	return ""
+}
+
+func PathIndexPointer(s string, i Index) *string {
+	v := PathIndex(s, i)
+	if v == "" {
+		return nil
+	}
+	return &v
+}
+
+func SplitN(s string, n int) ([]string, bool) {
+	parts := strings.Split(s, "/")
+	return parts, len(parts) == n
+}
+
+// Copy
+// Plugin framework doesn't support embedded structs.
+// https://github.com/hashicorp/terraform-plugin-framework/issues/242
+// We can't share the same functions for resource and datasource by default.
+// To do so, we design all functions for resource and copy it to datasource.
+func Copy(from, dest any) func() {
```

Comment:
nit: prefer specific types over `any` if possible; if not, document why `any` is necessary